### PR TITLE
Spice Flight: v2 Design

### PR DESCRIPTION
**Pull request by brianrisk**
@moollaza @thm Changes with this PR:

1) fluid layout.  This was done with tables as this is tabular data (columns represent a data type)

2) progress bar and position dot have been added:

screen shot 2015-11-23 at 2 08 31 pm

3) delayed flight styling applied:

screen shot 2015-11-23 at 1 26 00 pm

4) Flights that have yet to depart list the time until their departing times:

screen shot 2015-11-23 at 1 52 38 pm

5) If a scheduled flight arrival has passed, but the data has yet to confirm landing, this is listed as "scheduled arrival":

screen shot 2015-11-23 at 2 01 55 pm

-----
IA Page: https://duck.co/ia/view/route

### CODE REVIEW

**Comment by moollaza** [comment]
@brianrisk nicely done! It looks like there's a merge conflict though. I think you'll need to pull in master

**Comment by moollaza** [comment]
Looks like there were some changes to `flights_route` on master that are conflicting

**Comment by thm** [comment]
@brianrisk awesome, this is perfect! near-flawless execution. I've attached a screenshot outlining some minor issues. In addition to that, can you also lower the font-size of the column titles by 2px? (so that's for "departure", "scheduled", "gate" etc)
feedback

Other than that, I would love to see this live! cc @moollaza

File: share/spice/flights/route/content.handlebars [hunk 1]
```diff
@@ -1,27 +1,56 @@
...
+
+<div class="tile__departure">
+	<table>
+	<tr class="tx-clr--lt">
+		<td>DEPARTURE</td>
+		<td>SCHEDULED</td>
+		<td>GATE</td>
+		<td>TERMINAL</td>
+	</tr>
+	<tr class="text--primary">
```
Comment (moollaza) [inline share/spice/flights/route/content.handlebars]: We **should** be able to use our `tx--*` classes here to specify font sizes.

```html
<tr class="text--primary tx--16"> 
```

> More info: https://duckduckgo.com/styleguide#txt-n-color

File: share/spice/flights/route/flights_route.css [hunk 2]
```diff
@@ -1,5 +1,23 @@
 .tile--route  {
-    width: 37em;
+    padding: 1.44em;
+}
+
+.zci--route .inactive  {
+    background-color: #f7f7f7;
```
Comment (moollaza) [inline share/spice/flights/route/flights_route.css]: We'll need to use our `bg--*` classes here so this looks good in the dark theme as well. The classes swap the colours for the dark themes automatically.

@brianrisk how does it currently look in the dark theme?

/cc @thm we don't have a color class for `#e4e4e4` though, but we have `bg--platinum-light` which is `#e5e5e5`

duckduckgo_style_guide

**Comment by brianrisk** [comment]
@thm The dot alignment is tough because due to the fact that in the eyes of CSS it has no width, the center of the dot is at 100%.  I'd love to say something like (100% - 4px).  However, what I've done is to scale all in-progress flights to have a max of 98% and if it has landed, the dot is not present and the bar is at 100%.  Here's the effect:

screen shot 2015-11-24 at 12 30 24 pm

I'm noticing that flights that landed hours ago are often dominating the display.  I've seen on the Sports IA that the tiles are advanced to the first active game.  How can I do that here?

screen shot 2015-11-24 at 12 24 13 pm

**Comment by moollaza** [comment]
>I'm noticing that flights that landed hours ago are often dominating the display. I've seen on the Sports IA that the tiles are advanced to the first active game. How can I do that here?

I think this is because we widened the window of flights to 24 hours +/- from the current time to get more results.

You should be able to set `meta.selectedItem` to set the result we should scroll to:

```js
meta: {
  sourceName: "...",
  sourceUrl: "...",
  selectedItem: "<array index of tile>"
}
```

**Comment by brianrisk** [comment]
Display with dark and light themes:

screen shot 2015-11-24 at 3 18 46 pm
screen shot 2015-11-24 at 3 19 00 pm

**Comment by brianrisk** [comment]
@moollaza I'm setting selectedItem to the first non-landed index (0 of all are landed), but the result is that the tiles do not advance and the normally darkened "landed" tiles now have this weird lighter background:

screen shot 2015-11-24 at 3 49 51 pm

Am getting this JS error, but it looks like it's coming from codio?

screen shot 2015-11-24 at 3 51 49 pm

**Comment by moollaza** [comment]
> I'm setting selectedItem to the first non-landed index (0 of all are landed), but the result is that the tiles do not advance and the normally darkened "landed" tiles now have this weird lighter background:

@brianrisk it looks like there may be a bug in this code right now -- or at least I can't get it to work myself either, so we'll need to add the scroll-to functionality later

**Comment by brianrisk** [comment]
@moollaza  I've commented the problematic property.  @thm I feel this is ready for a design review.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=Jetblue Boston to Los Angeles)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/route_mobile.png)

**Comment by moollaza** [comment]
@brianrisk @thm this is now live on Beta:  https://beta.duckduckgo.com/?q=Jetblue%20Boston%20to%20Los%20Angeles

It looks like the the "Delayed - Scheduled arrival X minutes ago" is a little too long, causing the text to wrap and the height to change. 

@thm how would you like that to be handled? Should the tile stretch to fit?

**Comment by moollaza** [comment]
jetblue_boston_to_los_angeles_at_duckduckgo

**Comment by moollaza** [comment]
@brianrisk also, looks like for some reason the first tile on mobile isn't filling the width, but the others do: 
screen shot 2015-11-24 at 21 39 56

**Comment by thm** [comment]
> It looks like the the "Delayed - Scheduled arrival X minutes ago" is a little too long, causing the text to wrap and the height to change ... how would you like that to be handled? Should the tile stretch to fit?

In the original design the timestamps were shortened. That should buy us some space: 
— `2 hours` should be `2h`
— `1 hour 23 minutes` should be `1h 23m`
— `an hour ago` should be `1h ago`
— `Scheduled arrival 31 minutes ago` should be `Scheduled arrival 31m ago`

@brianrisk Is it possible to somehow get an estimated arrival time (eg. `+32m`)? because that can make the last item in the list above even shorter, and it'll be great for the ux. I also think we can make the scheduled arrival column `$red` for delayed flights like in the original design:
screen shot 2015-11-25 at 12 30 56 pm

> how should the mobile UI look?

@moollaza Can we just keep the carousel view and have the tiles retain their dynamic width? If so, then we don't need any special changes for mobile.

**Comment by moollaza** [comment]
>@moollaza Can we just keep the carousel view and have the tiles retain their dynamic width? If so, then we don't need any special changes for mobile.

@thm as mentioned on Asana, it's not currently possible to force a mobile carousel so for now, we'll have to force 100% width tiles on mobile, which should be fairly easy with out CSS classes:

E.g. `<div class="whole--screen-s">`

**Comment by brianrisk** [comment]
@thm How about I change the text to "Est. arrival" instead of "Sheduled arrival"?  

I feel like we should present one time, and that is the best estimate as to the actual arrival.  To me, the (+32m) notation is possibly confusing (is it 2 hours 11 minutes + 32 minutes = 2 hours 43 minutes, or does that mean we are 32 minutes past meaning it was originally scheduled to arrive in 1 hour 39 minutes?).

**Comment by thm** [comment]
@brianrisk Good point, I think "Est. arrival" is a good solution!

**Comment by thm** [comment]
> ... so for now, we'll have to force 100% width tiles on mobile, which should be fairly easy with out CSS classes

@moollaza I think that works. If we can align the text in the table columns to the first line, then I think the text wrapping isn't really a big deal. @brianrisk so, aligning "11:08 PM" with "Los Angeles, CA" in the screenshot below:
screen shot 2015-12-02 at 2 08 48 pm

**Comment by moollaza** [comment]
> @brianrisk so, aligning "11:08 PM" with "Los Angeles, CA" in the screenshot below:

@brianrisk @thm that's super easy to achieve -- we just need to set `vertical-align: top` in the CSS

It looks like this PR is very close to being finished, I'm super excited :+1:

**Comment by brianrisk** [comment]
@moollaza Awesome!  The css alignment is now in.  @thm how's it look to you?

**Comment by moollaza** [comment]
@brianrisk @thm changes are live on beta:

screen shot 2015-12-08 at 10 09 36
screen shot 2015-12-08 at 10 08 24

@brianrisk it looks like we're just missing the full-width tiles on mobile.

**Comment by brianrisk** [comment]
@moollaza home stretch.  Looks like I don't have access to the parent div ("tile--route") in handlebars.  Should I add the class="whole--screen-s" via JS?

**Comment by moollaza** [comment]
You should be able to apply it with `elClass: { tile: 'className' } `

--
Zaahir Moolla
https://duckduckgo.com

**Comment by brianrisk** [comment]
@moollaza Ok, that's in.

**Comment by moollaza** [comment]
@brianrisk @thm it's live -- It looks like we need to add `whole` to the `<table>` so that the table stretches to fit the tile:

dashboard

I think we may need to also add `whole--screen-m` to force the tiles to stretch for landscape mode?

On the iPhone 6 it's not wide enough for 2 side-by-side tiles:

dashboard

**Comment by brianrisk** [comment]
@moollaza is the class you mean to ad to the table "whole--screen-s" or simply "whole"?

**Comment by moollaza** [comment]
> @moollaza is the class you mean to ad to the table "whole--screen-s" or simply "whole"?

Just `whole` so that it always fills 100% of the tile width

**Comment by brianrisk** [comment]
@moollaza I've added the styling, but am flying blind as to if it is having the desired effect on the mobile display.

**Comment by moollaza** [comment]
@brianrisk I'm deploying the updates to beta.

@thm how should this look for landscape screens? Full width or half width?

**Comment by thm** [comment]
@moollaza @brianrisk looking at those screenshots you posted above, on iPhone 6/6+ and Nexus 6 kind of screens we can definitely show two tiles per row. On iPhone 4 sizes we should probably stick with 1 per row to avoid crammed content. Can we set a threshold at around iPhone 5 and larger to show 2 tiles per row?

**Comment by thm** [comment]
around >600px seems ideal to me

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=Jetblue Boston to Los Angeles)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/route_mobile.png)

**Comment by moollaza** [comment]
@thm hmm unfortunately it looks like mobile will be really hard to handle with a tile view, because based on the content of the tile, the text will wrap and then we have no way of knowing or setting the grid of tiles to the same height.

We may need to force full-width tiles (which granted looks odd past the 600px mark) or we need to force the text to a max of 1 or 2 lines and force the tiles to have the same height

Also, because of the padding on our tiles, our 50% class doesn't actually work, we'd need to force 48% width or remove the tile padding :frowning: 

Here's some examples:

american_airlines_jfk_to_yyz_at_duckduckgo

american_airlines_jfk_to_yyz_at_duckduckgo

-----

Alternatively, we can put this design on hold until we can force a mobile carousel view?

/cc @bsstoner

**Comment by bsstoner** [comment]
I'd much rather see us force the mobile carousel than hack the CSS. We already have a way to do it internally (see Places), it just needs to be exposed to Spice. I opened an asana task to do it, should be quick change.

**Comment by bsstoner** [comment]
Sorry, I should have updated this PR a while ago when the internal change went live. You can now pass `disableMobileGrid: true` in the meta block and it will force the tiles to stay in the horizontal carousel on mobile. I think we should be good to finish this off now?

**Comment by jagtalon** [comment]
Awesome, thanks @bsstoner!

**Comment by moollaza** [comment]
FYI I'm finishing this up...nearly done...

**Comment by argibbs** [comment]
Happened across this at random. Not sure how much you care, but timezones are a problem. For example, let's take your sample query "jetblue boston to los angeles". 

I'm based in London. There's a jetblue flight departing at 7:10am EST. Running your example query at 12 noon GMT (i.e. 7am EST) shows the flight as "On Time - Landed 2 hours ago". I'm guessing you generate that "Landed 2 hours ago" string using my local time as reported by the browser, but that leads to erroneous values.

**Comment by moollaza** [comment]
>  Not sure how much you care, but timezones are a problem

@argibbs thanks for pointing this out! and yes we ***definitely*** care!

I had noticed this when wrapping this PR up and was planning to fix it. Would you be interested in fixing it?

**Comment by argibbs** [comment]
Thanks for asking, but I don't have the experience sorry. And not enough hours in the day. I felt bad about the drive by bug report, but I happened upon it at random and timezone bugs are hard enough to spot that I thought I'd make an exception.

**Comment by moollaza** [comment]
Closing. Being continued in #2602

**Pull request closed without merging**